Docs/src/tutorials/basictutorials/code/basictutorial1fullsource.md

```diff
@@ -0,0 +1,151 @@
+# Basic Tutorial 1 - Full source code {#basictutorial1fullsource}
+
+## Header file
+
+```cpp
+
+This source file is a part of OGRE
+
+#ifndef SRC_TUTORIALAPPLICATION_HPP_
+#define SRC_TUTORIALAPPLICATION_HPP_
+
+#include <OGRE/Ogre.h>
+#include <OGRE/Bites/OgreApplicationContext.h>
+#include <OGRE/Bites/OgreInput.h>
+#include <OGRE/RTShaderSystem/OgreRTShaderSystem.h>
+#include <OGRE/Bites/OgreApplicationContext.h>
+
+class TutorialApplication
+        : public OgreBites::ApplicationContext
+        , public OgreBites::InputListener
+{
+public:
+    TutorialApplication();
+    virtual ~TutorialApplication();
+
+    void setup();
+    bool keyPressed(const OgreBites::KeyboardEvent& evt);
+};
+
+#endif /* SRC_TUTORIALAPPLICATION_HPP_ */
+
+
+```
+
+## Source file
+
+```cpp
```

Comment:
could you instead of pasting the c++ source into the doc, reference a compilable source file using the @snippet command?

see:
- https://github.com/OGRECave/ogre/blob/master/Docs/src/tutorials/setup.md
- https://github.com/OGRECave/ogre/blob/master/Samples/Bootstrap/main.cpp

store the sources as `Samples/Tutorials/basic1.cpp` etc., if you can also provide a CMakeLists to compile them.

This way we can ensure that the tutorials are up to date - even if the API changes.

Reply:
Making Cmake project for tutorials is good idea! 
The question is what structure of the project should I make?

For the first glance it could be main CMake file and then set of small child cmake projects. But in this case main.cpp will be copy and pasted for every project. 

I see the project structure as this:

<pre>
CMakeLists.txt
+basic1
   - CMakeLists.txt
   - main.cpp
   - BasicTutorial1.cpp
   - BasicTutorial1.hpp
+basic2
   - CMakeLists.txt
   - main.cpp
   - BasicTutorial1.cpp
   - BasicTutorial1.hpp
<pre>

In this case we can have a set of targets which will be compiled all at once.

Reply:
as the point of the tutorials are the Ogre concepts and not c++ best practices, I would just put everything in one .cpp file (again see Bootstrap sample) as:
```
CMakeLists.txt
- add_executable(BasicTutorial1 BasicTutorial1.cpp)
- ...
- add_executable(BasicTutorialN BasicTutorialN.cpp)

BasicTutorial1.cpp
...
BasicTutorialN.cpp
```

Reply:
Yeah. That is what I thought later would be better in this case.
Will do it in this way.